Rework kbd-intr handler code and fix pending exceptions when disabling kbd-intr char

dpgeorge opened the pull request:
There are 2 commits here:
- rework and factor existing keyboard-interrupt handler code to move it from `lib/utils/interrupt_char.[ch]` to the py-core (because it's really tightly coupled to the scheduler)
- fix issues #4921, #5469 and #5488 by checking for pending exceptions (and dealing with them) when disabling the ctrl-C character

dpgeorge committed:
all: Move interrupt char handling code from lib/utils to py/ core.

It's tightly coupled to the scheduler and VM so should be part of the core.
This also simplifies ports usage of this ctrl-C feature.

dpgeorge committed:
py/scheduler: Handle pending exceptions when disabling kbd interrupts.

Pending exceptions would otherwise be handled later on where there may not
be an NLR handler in place.

Fixes issues #4921 and #5488.

dpgeorge committed:
unix, windows: Fix windows build by renaming hal function.

dpgeorge committed:
unix, windows: Call core mp_set_interrupt_char in HAL version.

dpgeorge committed:
tests/unix: Add coverage test for interrupt char handling.

dpgeorge committed:
windows: Include py/runtime.h for mp_set_interrupt_char.

dlech commented on py/runtime.c:
```diff
@@ -73,6 +73,8 @@ void mp_init(void) {
 
     #if MICROPY_KBD_EXCEPTION
+    // Disable intercepting any interrupt character
+    MP_STATE_VM(interrupt_char) = -1;
```
Should this be `MP_INTERRUPT_CHAR_DISABLE_AND_CANCEL_PENDING` instead of `-1`?

dpgeorge replied:
Yes indeed!

dpgeorge replied:
Now fixed.

dlech commented on ports/stm32/usbd_cdc_interface.c:
```diff
@@ -267,7 +266,7 @@ int8_t usbd_cdc_receive(usbd_cdc_state_t *cdc_in, size_t len) {
     // copy the incoming data into the circular buffer
     for (const uint8_t *src = cdc->rx_packet_buf, *top = cdc->rx_packet_buf + len; src < top; ++src) {
-        if (cdc->attached_to_repl && mp_interrupt_char != -1 && *src == mp_interrupt_char) {
+        if (cdc->attached_to_repl && MP_STATE_VM(interrupt_char) != -1 && *src == MP_STATE_VM(interrupt_char)) {
```
Should this be `MP_INTERRUPT_CHAR_DISABLE_AND_CANCEL_PENDING` instead of `-1`?

dpgeorge replied:
Hmm, it should actually test for `<0` because now `interrupt_char` could be -1 or -2... maybe it could do with a macro like `MP_INTERRUPT_CHAR_ENABLED()`.

dpgeorge replied:
Now fixed.  I just removed this part of the if-logic, it's not needed.

mbuesch commented:
I can confirm that this fixes #5469.
Thanks!

dpgeorge commented:
> I can confirm that this fixes #5469.

Great, thanks for testing!

dpgeorge committed:
py/runtime: Fix initialisation of interrupt_char.

dpgeorge committed:
stm32/usbd_cdc_interface: Remove "interrupt_char != -1" check.

It's not needed.  The C integer implicit promotion rules mean that the
uint8_t of the incoming character and the int16_t of interrupt_char are
both promoted to a (signed) int.  Thus the uint8_t incoming character can
never be equal to -1 or -2 (the values of interrupt_char that indicate that
interruption is disabled).

dpgeorge committed:
samd/mphalport: Convert all negative interrupt chars to -1 for TinyUSB.

dpgeorge committed:
py/scheduler: Move clearing of kbd tracebace to mp_keyboard_interrupt.

dpgeorge committed:
py/schedule: Fix new mp_set_interrupt_char to be atomic.

dpgeorge commented:
Superseded by #5610
